wcfsetup/install/files/lib/action/FileDownloadAction.class.php

```diff
@@ -0,0 +1,118 @@
+<?php
+
+namespace wcf\action;
+
+use Laminas\Diactoros\Response;
+use Laminas\Diactoros\Response\EmptyResponse;
+use Laminas\Diactoros\Stream;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use wcf\data\file\File;
+use wcf\http\ContentDisposition;
+use wcf\http\Helper;
+use wcf\system\exception\IllegalLinkException;
+use wcf\system\exception\PermissionDeniedException;
+use wcf\util\FileUtil;
+
+/**
+ * Offers a file for download.
+ *
+ * @author Alexander Ebert
+ * @copyright 2001-2024 WoltLab GmbH
+ * @license GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @since 6.1
+ */
+final class FileDownloadAction implements RequestHandlerInterface
+{
+    public function handle(ServerRequestInterface $request): ResponseInterface
+    {
+        $parameters = Helper::mapQueryParameters(
+            $request->getQueryParams(),
+            <<<'EOT'
+                array {
+                    id: positive-int
+                }
+                EOT,
+        );
+
+        $file = new File($parameters['id']);
+        if (!$file->fileID) {
+            throw new IllegalLinkException();
+        }
+
+        $processor = $file->getProcessor();
+        if ($processor === null) {
+            throw new IllegalLinkException();
+        }
+
+        if (!$processor->canDownload($file)) {
+            throw new PermissionDeniedException();
+        }
+
+        $eTag = \sprintf(
+            '"%d-%s"',
+            $file->fileID,
+            \substr($file->fileHash, 0, 8),
+        );
+
+        $httpIfNoneMatch = $_SERVER['HTTP_IF_NONE_MATCH'] ?? '';
```

Comment:
- At a minimum this should use `$request->getHeaderLine()`, instead of `$_SERVER`.
- A more correct solution would be leveraging Guzzle’s `Header::splitList()` + an in_array check (see the example in the README: https://github.com/guzzle/psr7?tab=readme-ov-file#guzzlehttppsr7headersplitlist). 
- The most correct solution would be actually using the “weak comparison function” to compare two ETags, as per [RFC 9110#13.1.2](https://www.rfc-editor.org/rfc/rfc9110#name-if-none-match). Basically an ETag would be a match for purposes of `if-none-match`, if it is identical after stripping the `W/` prefix from both ETags.

The last bullet point would be relevant if the web server gzips the attachment, thus turning the strong ETag by WoltLab Suite into a weak one.

PS: Technically there's also the asterisk to handle :smile: